File: Mission_to_MARS_Java/app/src/main/java/com/example/mission_to_mars/Simulation.java
package com.example.mission_to_mars;

import android.content.Context;
import android.content.res.AssetManager;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Scanner;

public class Simulation {
    private Context context;

    public Simulation(Context context) {
        this.context = context;
    }

    public ArrayList<Item> loadItems(String fileName) throws IOException{
        AssetManager assManeger = context.getAssets();
        ArrayList<Item> loadManifest = new ArrayList();

        Scanner filescanner = new Scanner(assManeger.open(fileName));

        while (filescanner.hasNextLine()) {

            Item newItem = new Item();
            String[] tokens = filescanner.nextLine().split("=");
            newItem.name = tokens[0];
            newItem.weight = Integer.parseInt(tokens[tokens.length-1]);

            loadManifest.add(newItem);
        }

        return loadManifest;
    }

    public ArrayList<Rocket> loadU1(ArrayList<Item> list) {
        ArrayList<Rocket> rocketListOne = new ArrayList<>();
        Rocket rocketOne = new U1();

        int itemCounter = 1;
        int rocketCounter = 1;

        for (Item i : list) {

            while (!rocketOne.canCarry(i)) {

                rocketCounter++;
                rocketListOne.add(rocketOne);
                rocketOne = new U1();
            }
            rocketOne.carry(i);

            itemCounter++;
        }
        rocketListOne.add(rocketOne);

        return rocketListOne;
    }

    public ArrayList<Rocket> loadU2(ArrayList<Item> list) {

        ArrayList<Rocket> rocketListTwo = new ArrayList();
        Rocket rocketTwo = new U2();

        int itemCounter = 1;
        int rocketCounter = 1;

        for (Item i : list) {

            while (!rocketTwo.canCarry(i)) {

                rocketCounter++;
                rocketListTwo.add(rocketTwo);
                rocketTwo = new U1();
            }
            rocketTwo.carry(i);

            itemCounter++;
        }
        rocketListTwo.add(rocketTwo);

        return rocketListTwo;
    }

    public int runSimulation(ArrayList<Rocket> rockets) {

        int totalCost = 0;

        for (Rocket currentRocket : rockets) {
            totalCost += currentRocket.getCost();

            do{
                totalCost += currentRocket.getCost();
            }while (!currentRocket.launch() || !currentRocket.land());
        }
        return totalCost;
    }

}

File: Mission_to_MARS_Kotlin/app/src/main/java/com/example/mission_to_mars_kotlin/Simulation.kt
package com.example.mission_to_mars_kotlin

import android.content.Context
import android.content.res.AssetManager
import java.io.IOException
import java.util.*
import kotlin.collections.ArrayList

class Simulation(var context: Context) {

    @Throws(IOException::class)
    fun loadItems(fileName: String): ArrayList<Item>{
        val assManeger: AssetManager = context!!.assets
        val loadManifest: ArrayList<Item> = ArrayList()

        val filescanner = Scanner(assManeger.open(fileName))

        while (filescanner.hasNextLine()) {
            val newItem: Item = Item()
            val tokens = filescanner.nextLine().split("=").toTypedArray()
            newItem.name = tokens[0]
            newItem.weight = tokens[tokens.size - 1].toInt()
            loadManifest.add(newItem)
        }

        return loadManifest
    }

    fun loadU1(list: ArrayList<Item>): ArrayList<Rocket>{
        val rocketListOne : ArrayList<Rocket> = ArrayList()
        var rocketOne: Rocket = U1()

        var itemCounter = 1
        var rocketCounter = 1

        for (i in list) {
            while (!rocketOne.canCarry(i)) {
                rocketCounter++
                rocketListOne.add(rocketOne)
                rocketOne = U1()
            }
            rocketOne.carry(i)
            itemCounter++
        }
        rocketListOne.add(rocketOne)

        return rocketListOne
    }

    fun loadU2(list: ArrayList<Item>): ArrayList<Rocket>{
        val rocketListTwo: ArrayList<Rocket> = ArrayList()
        var rocketTwo: Rocket = U2()

        var itemCounter = 1
        var rocketCounter = 1

        for (i in list) {
            while (!rocketTwo.canCarry(i)) {
                rocketCounter++
                rocketListTwo.add(rocketTwo)
                rocketTwo = U1()
            }
            rocketTwo.carry(i)
            itemCounter++
        }
        rocketListTwo.add(rocketTwo)

        return rocketListTwo
    }

    fun runSimulation(rockets: ArrayList<Rocket>): Int{
        var totalCost = 0

        for (currentRocket in rockets) {
            totalCost += currentRocket.getCost()

            do {
                totalCost += currentRocket.getCost()
            }while (!currentRocket.launch() || !currentRocket.land())
        }
        return totalCost
    }
}